sw/device/tests/kmac_endianess_test.c

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_kmac.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+/**
+ * KMAC test description.
+ */
+typedef struct kmac_test {
+  dif_kmac_mode_kmac_t mode;
+  dif_kmac_key_t key;
+
+  const char *message_big_endian;
+  const char *message_little_endian;
+  size_t message_len;
+
+  size_t digest_len;
+} kmac_test_t;
+
+/**
+ * KMAC test.
+ *
+ * Example taken from:
+ * https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/KMAC_samples.pdf
+ */
+const kmac_test_t kmac_test = {
+    // KMAC
+    .mode = kDifKmacModeKmacLen128,
+    .key =
+        (dif_kmac_key_t){
+            .share0 = {0x43424140, 0x47464544, 0x4B4A4948, 0x4F4E4D4C,
+                       0x53525150, 0x57565554, 0x5B5A5958, 0x5F5E5D5C},
+            .share1 = {0},
+            .length = kDifKmacKeyLen256,
+        },
+    .message_big_endian = "\x00\x01\x02\x03",
+    .message_little_endian = "\x03\x02\x01\x00",
+    .message_len = 4,
+    .digest_len = 8,
+};
+
+/**
+ * Big to little endianess converter.
+ *
+ * Helper function to convert from bit to little endian.
+ *
+ * @param big_endian the number in big endian representation.
+ * @return the number in little endian representation.
+ */
+uint32_t big_to_little_endian(uint32_t big_endian) {
+  uint32_t little_endian = 0;
+  little_endian = (big_endian >> 24) & 0xff;
+  little_endian |= (big_endian >> 8) & 0xff00;
+  little_endian |= (big_endian << 8) & 0xff0000;
+  little_endian |= (big_endian << 24) & 0xff000000;
+
+  return little_endian;
+}
+
+/**
+ * State endianess test.
+ *
+ * This test runs a KMAC operation twice, in the first run with big endian for
+ * the state output and in the second run with little endian. The big endian
+ * state output is converted to little endian and a comparison is conducted.
+ *
+ * @return OK or error.
+ */
+status_t test_state_endianess(void) {
+  uint32_t out_big_endian[kmac_test.digest_len];
+  uint32_t out_little_endian[kmac_test.digest_len];
+  uint32_t big_to_little[kmac_test.digest_len];
+
+  // Intialize KMAC hardware.
+  dif_kmac_t kmac;
+  dif_kmac_operation_state_t kmac_operation_state;
+  CHECK_DIF_OK(
+      dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR), &kmac));
+
+  LOG_INFO("Testing big endian state.");
+  dif_kmac_config_t config = (dif_kmac_config_t){
+      .entropy_mode = kDifKmacEntropyModeSoftware,
+      .entropy_seed = {0xb153e3fe, 0x09596819, 0x3e85a6e8, 0xb6dcdaba,
+                       0x50dc409c, 0x11e1ebd1},
+      .entropy_fast_process = kDifToggleEnabled,
+      .output_big_endian = true,
```

Comment:
Should you also set the `message_big_endian` to true? You're using the big endian messages for this test but I think if you don't configure the corresponding struct member to true, it might actually default to false.